### PR TITLE
[IMP] web_widget_x2many_2d_matrix: New option field_att_

### PR DESCRIPTION
**Pull request by pedrobaeza**
Adding this option to the widget, you can define a field in the x2many
model that is evaluated for enabling or not the field in the matrix.
With this, you have a way to control if some matrix combinations are
not allowed.

NOTE: This doesn't prevent to require to fill the full matrix with
all the combination records.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
@hbrunn your review is welcome.

cc @Tecnativa

File: web_widget_x2many_2d_matrix/README.rst [hunk 1]
```diff
@@ -50,15 +53,25 @@ field_label_y_axis
     Use another field to display in the table header
 field_value
     Show this field as value
+field_editability
+    Use this field to evaluate if a cell should be editable or not. It
```
Comment (hbrunn) [inline web_widget_x2many_2d_matrix/README.rst]: finish the sentence, probably explaining that this is supposed to be truthy if editable, and not truthy otherwise. How about a serialized field that returns a dict with properties such as `readonly` and `invisible`?

File: web_widget_x2many_2d_matrix/static/src/xml/web_widget_x2many_2d_matrix.xml [hunk 2]
```diff
@@ -16,7 +16,13 @@
                         <th><t t-esc="widget.get_y_axis_label(y)" /></th>
                         <td t-foreach="widget.get_x_axis_values()" t-as="x" t-att-class="'oe_list_field_cell' + (widget.is_numeric ? ' oe_number' : '')" t-att-data-x="x">
                             <span t-att-class="widget.get_xy_value_class()">
-                                <input class="edit" t-att-data-id="widget.get_xy_id(x, y)" t-att-value="widget.format_xy_value(widget.get_xy_value(x, y))" />
+                                <t t-set="editable" t-value="widget.get_xy_editability(x, y)"/>
```
Comment (hbrunn) [inline web_widget_x2many_2d_matrix/static/src/xml/web_widget_x2many_2d_matrix.xml]: you could also say `t-attr-disabled="not widget.get_xy_editability(x, y) or None"` (an attribute with value `None` won't be set) - this way, you don't need to duplicate the code for the input

File: web_widget_x2many_2d_matrix/README.rst (hunk 1 above)
Comment (hbrunn) [inline web_widget_x2many_2d_matrix/README.rst]: now reading your actual implementation, how about a dict that is simply merged into the attributes using `t-att="widget.get_xy_editability(x, y)"`? Then it might make sense to rename the field, but I envision a lot of cool crazy stuff people can do with this

Comment (pedrobaeza) [reply web_widget_x2many_2d_matrix/README.rst]: I don't get you. Can you explain what you want to do?

File: web_widget_x2many_2d_matrix/static/src/xml/web_widget_x2many_2d_matrix.xml (hunk 2 above)
Comment (pedrobaeza) [reply web_widget_x2many_2d_matrix/static/src/xml/web_widget_x2many_2d_matrix.xml]: I think I tried that, but this is a special attribute that should be present or not. It's not possible to put `None`. Let me try again...

Comment (hbrunn) [reply web_widget_x2many_2d_matrix/static/src/xml/web_widget_x2many_2d_matrix.xml]: that's the trick with `None`. Either that or `False` (don't remember which one) causes the attribute not to exist at all. So the  function should return either `'disabled'` or `None`

File: web_widget_x2many_2d_matrix/README.rst (hunk 1 above)
Comment (hbrunn) [reply web_widget_x2many_2d_matrix/README.rst]: imagine I can return

``` python
{
    'disabled': 'disabled', # that's what you need
    'class': 'some class', # we can add some fancyness per field
    'onwhatever': 'injected javascript', # might be helpful in some use cases
    'style': 'display: none', # or just hide it?
    'out of ideas': 'but theres probably more cool stuff to do with this',
}
```

and all keys become attributes of the input. Very flexible.

File: web_widget_x2many_2d_matrix/static/src/xml/web_widget_x2many_2d_matrix.xml (hunk 2 above)
Comment (pedrobaeza) [reply web_widget_x2many_2d_matrix/static/src/xml/web_widget_x2many_2d_matrix.xml]: This is a JS expression, but you're right, doing this works: `t-att-disabled="widget.get_xy_editability(x, y) ? undefined : 'disabled'"`

**Comment by pedrobaeza** [comment]
@hbrunn I have implemented your suggestion and I'm improving slowly my JS skills. I have to implement an extra field in my TransientModel for computing `not bool(product_id)`, but it works. I have squashed directly because I'm starting the migration of the module and I need to cherry-pick this directly.

File: web_widget_x2many_2d_matrix/static/src/js/web_widget_x2many_2d_matrix.js [hunk 3]
```diff
@@ -38,12 +21,16 @@ openerp.web_widget_x2many_2d_matrix = function(instance)
         field_y_axis: 'y',
         field_label_y_axis: 'y',
         field_value: 'value',
+        // Putting this field by default, we always ensure editability
+        field_editability: 'id',
```
Comment (hbrunn) [inline web_widget_x2many_2d_matrix/static/src/js/web_widget_x2many_2d_matrix.js]: this one is obsolete then, right?

File: web_widget_x2many_2d_matrix/static/src/js/web_widget_x2many_2d_matrix.js [hunk 4]
```diff
@@ -261,6 +254,26 @@ openerp.web_widget_x2many_2d_matrix = function(instance)
...
+                    this.by_x_axis[x][y], this.fields_att[att]);
+                // Discard empty values
+                if (val) {
+                    vals[att] = val;
+                }
+            }
+            return vals;
+        },
+
+        get_xy_editability: function(x, y)
```
Comment (hbrunn) [inline web_widget_x2many_2d_matrix/static/src/js/web_widget_x2many_2d_matrix.js]: obsolete too?

File: web_widget_x2many_2d_matrix/static/src/js/web_widget_x2many_2d_matrix.js [hunk 5]
```diff
@@ -261,6 +254,26 @@ openerp.web_widget_x2many_2d_matrix = function(instance)
             return this.by_x_axis[x][y]['id'];
         },
 
+        get_xy_att: function(x, y)
+        {
+            vals = {};
```
Comment (hbrunn) [inline web_widget_x2many_2d_matrix/static/src/js/web_widget_x2many_2d_matrix.js]: in js, you should declare variables with the var keyword. Otherwise, you'll pollute the global namespace which can lead to weird and hard to debug interactions

**Comment by hbrunn** [comment]
this is great! A few recommendations, but otherwise :+1:

**Comment by pedrobaeza** [comment]
@hbrunn fixed!

**Review by lasley — APPROVED** [review]
TIL this module, very neat!

Changes look good, code and functional 👍

**Comment by elicoidal** [comment]
:+1:

**Comment by pedrobaeza** [comment]
Can you please merge as it's approved?

**Comment by pedrobaeza** [comment]
Thanks!
